### PR TITLE
[qt] bumpfee: offer user to reduce output for transactions without change

### PR DESCRIPTION
**Pull request by Sjors**
Builds on top of #12096. I'll rebase when that's merged.

If a user sends funds to a single destination, they might be trying to move all their funds to another wallet or an exchange. However it's also possible that they are making a payment and coin selection found a matching input.

With the former it's fine to reduce the output in order to increase the fee, but with the latter it's not, e.g. it could lead to underpaying an invoice. In that case we show the following warning:

schermafbeelding 2018-02-28 om 13 38 14

At the risk of too much UI clutter, the message could be made stronger by showing the original amount(s) vs. the new amount(s).

For comparison, this message is shown when a change output can be reduced:

schermafbeelding 2018-02-28 om 13 50 20

If in the future we're able to add inputs, we should probably offer the user a choice if they prefer to reduce an output instead. It's worth keeping that in mind when choosing the right wording here.

### CODE REVIEW

File: src/qt/walletmodel.cpp [hunk 1]
```diff
@@ -671,14 +671,30 @@ bool WalletModel::bumpFee(uint256 hash)
...
+
+    if (feebumper::CreateTransaction(wallet, hash, coin_control, 0 /* totalFee */, reduce_output, errors, old_fee, new_fee, mtx) != feebumper::Result::OK) {
         QMessageBox::critical(0, tr("Fee bump error"), tr("Increasing transaction fee failed") + "<br />(" +
             (errors.size() ? QString::fromStdString(errors[0]) : "") +")");
          return false;
     }
 
     // allow a user based fee verification
-    QString questionString = tr("Do you want to increase the fee?");
+    QString questionString;
```
Comment (promag) [inline src/qt/walletmodel.cpp]: Nit, and the diff is smaller:
```cpp
QString questionString = tr(...);
if (reduce_output != -1) {
     questionString = tr(...);
}

**Review by promag — COMMENTED** [review]
This logic in `WalletModel::bumpFee` sounds misplaced. I would add a parameter to `feebumper::CreateTransaction` to return whether a non-change output was reduced and move the logic there.

**Comment by Sjors** [comment]
> add a parameter to feebumper::CreateTransaction to return whether a non-change output was reduced

@kallewoof can that be part of #12096, e.g. if `reduce_output==-2` it picks the biggest output?

File: src/qt/walletmodel.cpp [hunk 2]
```diff
@@ -671,14 +671,30 @@ bool WalletModel::bumpFee(uint256 hash)
     CAmount old_fee;
     CAmount new_fee;
     CMutableTransaction mtx;
-    if (feebumper::CreateTransaction(wallet, hash, coin_control, 0 /* totalFee */, errors, old_fee, new_fee, mtx) != feebumper::Result::OK) {
+
+    // If the transaction has a single output, ask user if it's OK to reduce:
+    int32_t reduce_output = -1;
+    auto it = wallet->mapWallet.find(hash);
```
[the remark this answers is not in the document]
Comment (Sjors) [reply src/qt/walletmodel.cpp]: Is there good documentation about when locks are appropriate?Other than: https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#lockingmutex-usage-notes

[the remark this answers is not in the document]
Comment (promag) [reply src/qt/walletmodel.cpp]: `mapWallet` is protected by `cs_wallet`. If this is moved to `feebumper::CreateTransaction` then the lock is already held there.

**Comment by luke-jr** [comment]
I don't think this is the right route. While I think users should be able to reduce outputs if desired, we should by default offer to add inputs, not reduce outputs.

**Comment by promag** [comment]
@luke-jr I think that discussion is for #12096. Also see https://github.com/bitcoin/bitcoin/pull/12096#issuecomment-356089931.

**Comment by luke-jr** [comment]
Indeed, to rephrase: users should be able to reduce outputs; it just should require explicitly asking to do so, and shouldn't be the default/recommended behaviour.

**Comment by bedri** [comment]
If a customer paying for a good for the vendor and vendor agrees to pay the
transaction fee, then the customer can make a transaction with nearly zero
fee which will wait a long time for confirmations and then increase the fee
to the agreed amount and accepting the reduction in the output. So that
vendor will be able to pay the transaction fee.
If the vendor does not willing to pay the transaction fee (no agreement
made between the vendor and the customer about this) then the customer can
make a regular transaction to pay for the good. If customer decides that
he/she put a low fee on the transaction so that the transaction will take a
long time to occur then he/she should bumpfee without reducing the output.

So bumpfee reduction from the output should be optional and should be
asked. For the above showcase, it may also be a good idea that if a bumpfee
made by reducing the output, the receiving side of the transaction (vendor
in the above case) should also be asked if he/she agrees with the
reduction. If he/she does not agree then the bumpfee should not occur.

**Comment by Sjors** [comment]
@luke-jr `Cancel` is the default option, but without an alternative I can see how users would just click OK even if it's not a good idea. Perhaps it's better to hold back this PR until `feebumper::CreateTransaction` supports adding inputs. By default the `OK` button would just add an input. We could show each of the destination addresses with a check box "reduce amount". That would seem like a safer UX.

Unless that change takes forever. Maybe we can tag this with 0.17, and revisit the discussion if this better alternative isn't ready?

@bedri there's no way (that I know) for the wallet to communicate with the merchant. The only thing we can do is inform the user that they should do this.

**Pull request closed without merging**